### ISSUE TITLE
Is it "policy" that installing hub does not wrap git by default? Or...?

### ISSUE THREAD

**Issue by geoff-codes** [github/hub#1791]
I'm a bit confused, on a couple of issues.

Is it something like "policy" that _completions_ for git (wrapping hub) be installed by default, but that users **must** themselves make another, manual intervention in order to have this alias/wrapper enabled by default?

Or is this based on the assumption that doing so isn't possible without the installer or package manager editing either a _**user**_ or _**system**_ config file (which _is_ a bad idea)? Did prior versions of package manager formulae and/or install scripts do this, but this has been changed?

Because as I read https://hub.github.com, I should be able to `brew install hub`, after which I should be able to invoke `git version` and it will show me a hub-wrapped-git, but this definitely is not the case.

I ask because, while I agree that installers shouldn't make changes to user or root profile/config files, there **are** other, better, middle ground options these days. For example, `fish` "now" (I'm not actually sure how recent this addition is, but in any case) includes a "vendor functions" directory, as well as vendor completions; this would allow for a wrapping function to be added by default, without modifying config files.

I bringing this up not because I'm hugely invested in the convenience of not having to add a line to a config/profile file.

Its actually because I take serious issue with your **[very bold assertion that is somehow "safe"](https://github.com/github/hub/blob/master/README.md#aliasing)**.
[Let me know and I can split this out into a separate issue from the "vendor function" suggestion.]

Hear me out: sure, you're not overriding any underlying `git` functionality. But what if (in any of dozens of ways) something happens to the `hub` executable, or the user's `$PATH` is modified, or simply, another alias is added at some point. I'm not sure I even need to give you a proof of concept here:

```sh
alias git=hub
# Sometime later, some script is run, and either added to the same file, or by sourcing another...
alias hub='printf "\033[31mUh-oh!\033[0m\n"; command hub'
# And now try:
git version
```

With a vendor **function**, there are several advantages, in that:
- You can make sure `command hub` is being invoked, avoiding alias injection (and you could add code to `hub` itself making to sure its not being invoked by a wrapper that had path priority)
- Many package managers require at least a user to be in an admin group, if not sude/root, to add or modify these functions.
- But you can still give a standard user the ability to opt out, by, say, setting something HUB_DO_NOT_WRAP_GIT.
- A function (as opposed to simply an alias) can also include a fallback, so if (for instance) something happens to `hub` (`brew unlink hub`? During an upgrade?), `git` invocations will continue to function.

As an example (homebrew and fish), add this `git.fish` to `/usr/local/share/fish/vendor_functions.d`:
```fish
# git wrapper function for hub (includes fallback)

not command -sq hub; or set -q HUB_DO_NOT_WRAP_GIT
and functions -e git

command -sq hub; and not set -q HUB_DO_NOT_WRAP_GIT
and function git --wraps hub
  command -sq hub
  and command hub $argv
  or command git $argv
  return $status
end
```

So, let me know. Maybe I'm missing something. Maybe you disagree. Maybe you have a better solution. But if you like, I could have PRs ready for you and homebrew in a couple hours. If you wanted. (For the fish implementation, which could be used as a rubric; I'd let others take over for the other shells).

❦ Geoff

**Comment by geoff-codes** [comment]
P.S. @mislav I'm seeing now others have [brought up issues about the alias wrapper](https://github.com/github/hub/issues/153), but I'm honestly surprised no one has ever brought it up in terms of security, it seems?

**Comment by mislav** [comment]
> but that users **must** themselves make another, manual intervention in order to have this alias/wrapper enabled by default?

Yes. The alias is optional and opt-in. Furthermore, enabling the alias during installation process would require us to edit people's shell configuration files, which is a rabbit hole we don't want to go down in.

> Because as I read https://hub.github.com, I should be able to `brew install hub`, after which I should be able to invoke `git version` and it will show me a hub-wrapped-git

If you read it a bit more closely, you would have noticed a line that says `alias git=hub`.

> You can make sure `command hub` is being invoked, avoiding alias injection (and you could add code to `hub` itself making to sure its not being invoked by a wrapper that had path priority)

If I understand correctly, you're advocating for a function wrapper rather than a shell alias to achieve the same goal (hub wrapping git) but with more resilience and more security. What I don't understand is, why is hub supposed to guard against alias injection? On a compromised system where an attacker has the privilege to inject aliases into your interactive shell, they can literally alias anything to anything. How is having hub on the system with `alias git=hub` active in your shell making it more vulnerable?

**Comment by apjanke** [comment]
Maybe it would be useful if the Homebrew `hub` formula included a Caveat reminding the user to do the `alias git=hub` thing if they want it to be used by default.
